### DISCUSSION TITLE
check_grads fails to capture wrong gradients

### DISCUSSION THREAD

**Discussion by SNMS95** [jax-ml/jax#17542] [Show and tell]
I was trying to implement a simple bisection algorithm. I created the algorithm and a test function and it finds the root.
```
def find_root(
    f, x, lower_bound, upper_bound, tolerance=1e-12, max_iterations=128):
  for _ in range(max_iterations):
    y = 0.5 * (lower_bound + upper_bound)
    if upper_bound - lower_bound < tolerance:
      break
    if f(x, y) > 0:
      upper_bound = y
    else:
      lower_bound = y
  return y

# testing it
def sig_fn(design, b):
    return jax.nn.sigmoid(design + b).mean()  - 0.3

design = jax.random.uniform(jax.random.PRNGKey(0), (1000,))
root = find_root(sig_fn, design, -10., 10.)
root # -1.36 and it satisfies the equation!

# Test gradient with unrolling
def dummy_func(design):
    root = find_root(sig_fn, design, -10., 10.)
    new_des = jax.nn.sigmoid(design + root)
    return new_des.mean()

jax.grad(dummy_func)(design) # -> Gives reasonable numbers
``` 

However, I wanted to implement a `custom_jvp` so that there would be no unrolling and I made the following code
```
@partial(jax.custom_jvp, nondiff_argnums=(0, 2, 3, 4, 5))
def find_root():
    ...

@find_root.defjvp
def find_root_jvp(f, lower_bound, upper_bound, tolerance, max_iterations, primals, tangents):
    x,  = primals
    x_dot,  = tangents
    y = find_root(f, x, lower_bound, upper_bound, tolerance, max_iterations)
    grad_vals = jax.grad(f, argnums=(0, 1))(x, y)
    y_dot = np.dot(-(grad_vals[0] / grad_vals[1]), x_dot)
    return y, y_dot
```

When I check the gradients like before, it shows really small number (1e-20)! But when I do check_grads, no error pops up.
`check_grads(dummy_func, (design,),
                order=2, eps=1e-4, atol=1e-2)`
Why is this the case?
Is the second implementation wrong? if so, check_grads should have caught it right?

**Comment by jakevdp** [comment]
`atol` means "absolute tolerance", so any gradient difference smaller than `1E-2` is going to be considered equal. Since your true gradients are on order `1E-4` and your incorrect gradients are close to zero, `check_grads` is considering them equal. If you want a different check, you'll need a different `atol` value.

**Reply by SNMS95** [reply]
@jakevdp I was perhaps not clear in my explanation. 

The thing is that for the second implementation with the custom_jvp, check grads does not give any error but it produces gradients close to zero(which does not seem correct ).

For the unrolling case, even with float64 and an extremely tight tolerance, check grads always shows an error

**Reply by SNMS95** [reply]
I tried without specifying any atol and still the results remain the same; no assertion errors for the implicit differentiation and errors for the unrolling case

**Reply by jakevdp** [reply]
> check grads does not give any error but it produces gradients close to zero(which does not seem correct ).

My point is that if the correct gradients are smaller than `atol` (which it appears they are) then incorrect gradients near zero will not be considered incorrect, because the difference between the correct and incorrect gradients is smaller than `atol`

> I tried without specifying any atol and still the results remain the same

Note that the default `atol` is not zero, but some small nonzero value that depends on dtype: https://github.com/google/jax/blob/d0df18a76b8c3b261e6fd1f9d32bb0684a81456e/jax/_src/public_test_util.py#L74-L86 You might try setting `atol=0` explicitly if you're looking to validate gradients whose correct values are close to zero.
